### PR TITLE
[4.0] Dynamic exclusivity enforcement for non-escaping closures.

### PR DESCRIPTION
**Pull request by atrick**
Add support for dynamic exclusivity enforcement of non-escaping closures.

This is enabled in Swift 4 mode.

This triggers a runtime trap when exclusivity is violated in a way
that isn't apparent statically. For example, the non-escaping closure
passed to `performNoEscape` below requires a dynamic access check
before passing `x` as `inout`.

```
  func performNoEscape(_ closure: () -> ()) { closure() }
  
  func modifyAndPerform(_ var: inout Int, closure: () ->()) {
    var = 42
    closure()
  }

  var x = 3
  let c = { x = 7 } // x has escaped via an escaping closure.

  // The following non-escaping closure passes `x` inout.
  // This requires dynamic enforcement since `x` has already escaped.
  performNoEscape { modifyAndPerform(&x, closure: c) }
```
Tracked as:
<rdar://problem/32061282> [Exclusivity] Enforcement selection for noescape closure captures.

### CODE REVIEW

**Comment by atrick** [comment]
@rjmccall or @devincoughlin, here's a pile of commits that add up to dynamic enforcement of closures. They need to go through CCC for 4.0.

**Comment by atrick** [comment]
@swift-ci test.

**Comment by atrick** [comment]
@swift-ci test source compatibility.

**Comment by atrick** [comment]
@swift-ci test linux platform.

File: include/swift/SILOptimizer/Analysis/ClosureScope.h [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+  // with OptionalTransformRange. Unfortunately it exposes the internals of the
+  // analysis data.
+  struct IndexLookupFunc {
+    // A reference to all closure parent scopes ordered by their index.
+    const std::vector<SILFunction *> &indexedScopes;
+
+    IndexLookupFunc(const std::vector<SILFunction *> &indexedScopes)
+        : indexedScopes(indexedScopes) {}
+
+    Optional<SILFunction *> operator()(int &idx) const {
```
Comment (rjmccall) [inline include/swift/SILOptimizer/Analysis/ClosureScope.h]: This isn't a show-stopper, but why does this take idx by reference?

**Review by rjmccall — APPROVED** [review]
Looks good to me.  One question that doesn't affect semantics.

File: include/swift/SILOptimizer/Analysis/ClosureScope.h (hunk 1 above)
Comment (atrick) [inline include/swift/SILOptimizer/Analysis/ClosureScope.h]: Copy-paste oversight. I'll fix it on trunk now.
Is this a thumbs up for CCC?

**Comment by atrick** [comment]
@devincoughlin @rjmccall minor cleanups are on trunk:
https://github.com/apple/swift/pull/10415

Is this 4.0 PR good for CCC as is?

**Comment by atrick** [comment]
CCC Info:

Explanation: Dynamically enforce exclusivity in non-escaping closures.

Scope: This breaks any code that currently violates exclusivity, but can't be diagnosed statically. This is a rare occurrence.

Radar: rdar://problem/32061282 [Exclusivity] Enforcement selection for noescape closure captures.

Risk: (1) This introduces a considerable amount of code in the -Onone pipeline. That carries risk of introducing compiler assertions that can't easily be worked around (e.g. by disabling optimization). (2) There is a risk that some code that "happened to work" before in Swift 4 mode now traps at runtime (highly unlikely).

Testing: Unit tests have been added. (1) The compile-time risk will be addressed by building as much Swift code as is at our disposal. To this end, source compatibility tests pass. (2) The runtime risk, although unlikely to be a problem, will be addressed by encouraging Swift 4 projects to try a recent 4.0 build.

**Pull request merged**